### PR TITLE
dev: add docker healthcheck in compose

### PR DESCRIPTION
**Pull request by ftupas**
Time spent on this PR: 0.01

Resolves: #337, #371 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [X] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

Create a side-car container that does a healthcheck on the `starknet` service that it is ready to accept connections before deploying Kakarot contracts

# Does this introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -14,6 +14,22 @@ services:
     networks:
       - internal
 
+  healthcheck:
+    image: appropriate/curl  # A simple image that contains curl
```
Comment (ClementWalter) [inline docker-compose.yaml:18]: why not using the official image https://hub.docker.com/r/curlimages/curl ?

File: docker-compose.yaml [hunk 2]
```diff
@@ -14,6 +14,22 @@ services:
     networks:
       - internal
 
+  healthcheck:
```
Comment (ClementWalter) [inline docker-compose.yaml:17]: can we name this starknet-healthcheck?

**Review by ClementWalter — CHANGES_REQUESTED** [review]
lgtm, though I'd like to make sure that madara has no healthcheck route available first.

We should also add a healthcheck to katana, even though it's fast enough we actually want to make sure that we don't start the other containers until the "starknet" network is ready

File: docker-compose.yaml [hunk 3]
```diff
@@ -13,6 +13,9 @@ services:
       - "--dev"
     networks:
       - internal
+    healthcheck:
+      test: ["CMD", "echo", "ok"]
```
Comment (ClementWalter) [inline docker-compose.yaml:17]: can we just add a comment saying that this is a hack waiting for madara to implement something better?

Comment (ClementWalter) [reply docker-compose.yaml:17]: I think that Oak has added a HEALTHCHECK in the madara image, tbc but we could then remove this

Comment (ftupas) [reply docker-compose.yaml:17]: Indeed there is this pr https://github.com/keep-starknet-strange/madara/pull/915, I've asked Madara team to push the new image so we can just use that 👍

File: docker-compose.yaml [hunk 4]
```diff
@@ -2,7 +2,7 @@ version: "3.2"
 
 services:
   starknet:
-    image: ghcr.io/keep-starknet-strange/madara:v0.1.0-testnet-sharingan-beta.7.1.experimental.4
+    image: ghcr.io/keep-starknet-strange/madara:v0.1.0-beta
```
Comment (ftupas) [inline docker-compose.yaml:5]: Depends on this job https://github.com/keep-starknet-strange/madara/actions/runs/5677451093/job/15385784124

File: docker-compose.yaml [hunk 5]
```diff
@@ -17,16 +17,17 @@ services:
...
       - RPC_URL=http://starknet:9944
-      - EVM_PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 
+      - EVM_PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
     volumes:
       # Since STARKNET_NETWORK is not set, deployments will be saved without network folder
       - deployments:/app/kakarot/deployments
     depends_on:
-      - starknet
+      starknet:
+        condition: service_healthy
```
[the remark this answers is not in the document]
Comment (ftupas) [reply docker-compose.yaml:30]: Got merged https://github.com/dojoengine/dojo/pull/683 ✅

**Review by Eikix — COMMENTED** [review]
Since health check got merged in dojo, what now?:)

**Comment by ftupas** [comment]
> Since health check got merged in dojo, what now?:)

Let me pin the image tag for dojo 👍

File: docker-compose.katana.yaml [hunk 6]
```diff
@@ -3,7 +3,7 @@ version: "3.2"
 services:
   starknet:
     image: ghcr.io/dojoengine/dojo:v0.1.0
```
Comment (ftupas) [inline docker-compose.katana.yaml:5]: Will pin image once available https://github.com/dojoengine/dojo/pkgs/container/dojo

**Comment by Eikix** [comment]
Alright great

**Review by ClementWalter — DISMISSED** [review]

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Review by Eikix — APPROVED** [review]
lgtm

**Review by Eikix — DISMISSED** [review]
lgtm

File: docker-compose.yaml [hunk 7]
```diff
@@ -54,7 +57,7 @@ services:
       - KAKAROT_HTTP_RPC_ADDRESS=0.0.0.0:3030
       - STARKNET_NETWORK=http://starknet:9944
       - RUST_LOG=trace
-      - EVM_PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 
+      - EVM_PRIVATE_KEY=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
```
Comment (ClementWalter) [inline docker-compose.yaml:60]: I think that the RPC doesn't need this key

Comment (ftupas) [reply docker-compose.yaml]: You're right, tested it locally and worked without this, I've removed it 👍

File: docker-compose.katana.yaml [hunk 8]
```diff
@@ -2,8 +2,8 @@ version: "3.2"
 
 services:
   starknet:
-    image: ghcr.io/dojoengine/dojo:v0.1.0
-    command: 
+    image: ghcr.io/dojoengine/dojo:latest
```
Comment (ftupas) [inline docker-compose.katana.yaml:5]: This [pr](https://github.com/dojoengine/dojo/pull/685) got merged before the [latest nightly](https://github.com/dojoengine/dojo/pkgs/container/dojo) was released so it won't have the `curl` changes. For now I'll put latest here so when the new image is released, the `healthcheck` in Katana will work

File: Dockerfile [hunk 9]
```diff
@@ -30,6 +30,12 @@ EXPOSE 3030
 # this is required to have exposing ports work from docker, the default is not this.
 ENV KAKAROT_HTTP_RPC_ADDRESS="0.0.0.0:9545"
 
+# Add a health check to make sure the service is healthy
+HEALTHCHECK --interval=3s --timeout=5s --start-period=1s --retries=5 \
+  CMD curl --request POST \
+    --header "Content-Type: application/json" \
+    --data '{"jsonrpc": "2.0", "method": "eth_chainId", "id": 1}' http://${KAKAROT_HTTP_RPC_ADDRESS} || exit 1
+
```
Comment (ftupas) [inline Dockerfile:38]: I've added in this pr a `HEALTHCHECK` for the rpc

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**